lib/datadog/core/configuration/option.rb

```diff
@@ -38,22 +58,54 @@ def initialize(definition, context)
         # @param value [Object] the new value to be associated with this option
         # @param precedence [Precedence] from what precedence order this new value comes from
         def set(value, precedence: Precedence::PROGRAMMATIC)
-          # Cannot override higher precedence value
-          if precedence[0] < @precedence_set[0]
+          # Is there a higher precedence value set?
+          if @precedence_set > precedence
+            # This should be uncommon, as higher precedence values tend to
+            # happen later in the application lifecycle.
             Datadog.logger.info do
-              "Option '#{definition.name}' not changed to '#{value}' (precedence: #{precedence[1]}) because the higher " \
-                "precedence value '#{@value}' (precedence: #{@precedence_set[1]}) was already set."
+              "Option '#{definition.name}' not changed to '#{value}' (precedence: #{precedence.name}) because the higher " \
+                "precedence value '#{@value}' (precedence: #{@precedence_set.name}) was already set."
             end
 
+            # But if it happens, we have to store the lower precedence value `value`
+            # because it's possible to revert to it by `#unset`ting
+            # the existing, higher-precedence value.
+            # Effectively, we always store one value pre precedence.
```

Comment:
This is indeed a weird situation, but I think it makes the info message be a bit... misleading. Because it looks like the setting was ignored, but it's possible that it still may be acted on in the future.

So I would maybe suggest updating the message to mention this? Something along the lines of "Warning: Not changing option XXX to value YYY because it's being controlled via remote configuration. YYY will be used if remote configuration is disabled".

This makes the warning message less general than it currently is, but right now there shouldn't be any other way for this to happen, right? You can't really set a "default" after-the-fact, so the only option here seems to be programmatic being set after remote configuration. (And we can always add a spec that breaks if this assumption fails, so we don't forget to change something, if e.g. we add more precedence levels)